Adding TLS Proxy support

curiousjgeorge opened the pull request:
I need the ability to use various authentication mechanisms when connecting to a web proxy through HTTPS. This Pull Request improves support for connecting to a proxy via TLS/SSL. These libcurl options were made available in libcurl 7.52.0.

Options are configured via a ClientConfiguration object, for example:
```
clientConfig.proxyScheme = Aws::Http::Scheme::HTTPS;  // Enables TLS Proxy

// The following new fields are available for use:
clientConfig.proxySSLCertPath = "/path/example.crt";
clientConfig.proxySSLCertType = "PEM";
clientConfig.proxySSLKeyPath = "/path/example.key";
clientConfig.proxySSLKeyType = "PEM";
clientConfig.proxySSLKeyPassword = "hunter1";
```

I've ran a few tests on an internal TLS proxy to confirm my use case of using a Cert and Key with my TLS proxy. I also tried it with a key that required a passphrase. Not sure what I can do for public integration tests in this repo - if you have any pointers, please let me know.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

marcomagdy commented on aws-cpp-sdk-core/include/aws/core/client/ClientConfiguration.h:
```diff
@@ -122,6 +122,34 @@ namespace Aws
             /**
+            * Username to use for TLS Authentication with a proxy
+            */
+            Aws::String proxyTlsUsername;
```
I know we are far from consistent with naming to begin with, but at least we _try_ to stay consistent in the same class/file. So can you s/Tls/TLS and s/Ssl/SSL?

marcomagdy commented on aws-cpp-sdk-core/include/aws/core/client/ClientConfiguration.h:
```diff
@@ -122,6 +122,34 @@ namespace Aws
+            */
+            Aws::String proxySslKeyPath;
+            /**
+            * Type of private key file used to connect to an HTTPS proxy
```
Add an example or possible values, since it's hard to determine that from the type being a string.

marcomagdy commented on aws-cpp-sdk-core/include/aws/core/client/ClientConfiguration.h:
```diff
@@ -122,6 +122,34 @@ namespace Aws
+            */
+            Aws::String proxySslCertPath;
+            /**
+            * Type of proxy client SSL certificate
```
Add an example or possible values, since it's hard to determine that from the type being a string.

marcomagdy commented on aws-cpp-sdk-core/source/http/curl/CurlHttpClient.cpp:
```diff
@@ -427,6 +431,35 @@ void CurlHttpClient::MakeRequestInternal(HttpRequest& request,
                 curl_easy_setopt(connectionHandle, CURLOPT_PROXYPASSWORD, m_proxyPassword.c_str());
             }
+#if LIBCURL_VERSION_MAJOR >= 7
+#if LIBCURL_VERSION_MINOR >= 52
```
Your PR probably precedes this change, but can you follow the pattern used here:
https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/CMakeLists.txt#L94 and define `CURL_HAS_TLS_PROXY` and check that instead here.

curiousjgeorge commented:
Thanks for the review, @marcomagdy. I've updated the PR.

Additionally, I removed the `proxyTlsUsername` and `proxyTlsPassword` fields that were initially added in this PR. I couldn't get those working in my simple test.

marcomagdy commented:
LGTM, we'll merge it internally first to go through our CI and if all is good it will be pushed to the public repository a day or two later.
Thanks again.

marcomagdy commented:
One more thing. Could you comment on how you tested this? I'm exploring ways to add some CI tests in our pipeline. Would love to hear what you did to get a more comprehensive idea of my options.

curiousjgeorge commented:
The company I work at has a TLS web proxy that requires a certificate and key to use, so that greatly assisted my effort to test. The certificate and key are automatically generated for me and are both in the same file. Without a TLS web proxy that I could use, I'm not sure how I could have tested this.

My test was manual. I used some skeleton code to upload a file to s3 using the TLS proxy:

```
  Aws::SDKOptions options;
  Aws::InitAPI(options);
  Aws::Utils::Logging::InitializeAWSLogging(
    make_shared<Aws::Utils::Logging::DefaultLogSystem>(
      Aws::Utils::Logging::LogLevel::Trace,
      "aws_log_"));

  string filename = "test.txt";
  string keyname= "test.proxy.txt";
  string bucketname = "SCRUBBED";

  const string region = "SCRUBBED";
  string accessKey = "SCRUBBED";
  string secretKey = "SCRUBBED";


  // Set config to use a proxy
  string vip = "webproxy";
  int port = 8443;

  Aws::Client::ClientConfiguration clientConfig;
  clientConfig.region = region;
  clientConfig.scheme = Aws::Http::Scheme::HTTPS;
  clientConfig.proxyScheme = Aws::Http::Scheme::HTTPS;
  clientConfig.proxyHost = vip;
  clientConfig.proxyPort = port;

  clientConfig.proxySSLCertPath = "/var/credentials/user.pem";
  clientConfig.proxySSLKeyPath =  "/var/credentials/user.pem";
  //clientConfig.proxySSLCertType = "PEM";
  //clientConfig.proxySSLKeyType = "PEM";
  //clientConfig.proxySslKeyPassword = "password5";

  // Test connection with S3
  shared_ptr<Aws::Auth::SimpleAWSCredentialsProvider> creds = make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(accessKey, secretKey);
  Aws::S3::S3Client s3_client(creds, clientConfig);

  Aws::S3::Model::PutObjectRequest request;
  request.WithBucket(bucketname).WithKey(keyname);
  auto input_data = Aws::MakeShared<Aws::FStream>("PutObjectInputStream",
            filename.c_str(), std::ios_base::in | std::ios_base::binary);
  request.SetBody(input_data);

  auto outcome = s3_client.PutObject(request);

  if (outcome.IsSuccess())
  {
      std::cout << "Done!" << std::endl;
  }
  else
  {
      std::cout << "error: "
          << outcome.GetError().GetExceptionName() << std::endl
          << outcome.GetError().GetMessage() << std::endl;
  }
```

I manually tested the following:
1) TLS proxy worked when cert and key were in the same file
2) TLS proxy worked when cert and key were in different files
3) TLS proxy worked when proxySSLKeyType and proxySSLCertType were set to "PEM"
4) Ensure proxy failed when using an invalid certificate
5) Ensure proxy failed when using an expired certificate
6) To my key, I added a passphrase by running: `openssl pkcs8 -v1 PBE-SHA1-3DES -topk8 -in user.pem -out user.pem.pass`. Then I set the `proxySslKeyPassword` option (and updated the `proxySSLKeyPath` option) and ensured that the proxy worked.
7) Ensure proxy failed when using an invalid passphrase

curiousjgeorge commented:
Hi @marcomagdy: Any luck on getting this PR to pass your internal CI? Anything I can help with?

marcomagdy commented:
Yes, we merged it internally. It will be out with tomorrow's release.

marcomagdy commented:
This has been merged.
